src/EFCore.PG/Storage/ValueConversion/NpgsqlArrayConverter.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Diagnostics;
+using System.Linq.Expressions;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Storage.ValueConversion
+{
+    /// <summary>
+    /// A value converter that can convert between array types, given another <see cref="ValueConverter"/> for the
+    /// elements.
+    /// </summary>
+    public class NpgsqlArrayConverter<TModelArray, TProviderArray> : ValueConverter<TModelArray, TProviderArray>
+    {
+        public NpgsqlArrayConverter(ValueConverter elementConverter)
+            : base(ToProviderExpression(elementConverter), FromProviderExpression(elementConverter)) {}
```

Comment:
Given inappropriate type combinations, does the stack trace point to the type issue, or does it complain about a bad expression (e g. L48)? If it's the latter, it might be worth some type validation here.

Reply:
Good point, added some checks.